python/sdist/amici/splines.py

```diff
@@ -0,0 +1,1687 @@
+"""
+Splines
+-------
+This module provides helper functions for reading/writing splines with AMICI
+annotations from/to SBML files and for adding such splines to the AMICI C++
+code.
+"""
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from . import sbml_import
+    from typing import (
+        Sequence,
+        Union,
+        Tuple,
+        Optional,
+        Dict,
+        Any,
+        List,
+        Set,
+        Callable,
+    )
+
+    BClike = Union[
+        None, str,
+        Tuple[Union[None, str], Union[None, str]]
+    ]
+
+    NormalizedBC = Tuple[Union[None, str], Union[None, str]]
+
+import collections.abc
+import xml.etree.ElementTree as ET
+import numpy as np
+import sympy as sp
+import libsbml
+from abc import abstractmethod, ABC
+from itertools import count
+from sympy.core.parameters import evaluate
+from .import_utils import (
+    sbml_time_symbol,
+    amici_time_symbol,
+    annotation_namespace,
+)
+from .sbml_utils import (
+    pretty_xml,
+    mathml2sympy,
+    sbml_mathml,
+    get_sbml_units,
+    add_parameter,
+    add_assignment_rule,
+    SbmlAnnotationError,
+)
+from numbers import Integral
+
+
+def sympify_noeval(x):
+    with evaluate(False):
+        return sp.sympify(x)
+
+
+###############################################################################
+
+
+class UniformGrid(collections.abc.Sequence):
+    """
+    A grid of uniformly-spaced real points, computed with rational arithmetic.
+
+    Implements the :py:class:`collections.abc.Sequence` interface and can be
+    converted to a :py:class:`numpy.ndarray` (conversion to float can be
+    specified with ``dtype=float``).
+
+    TODO: document attributes
+    """
+
+    def __init__(
+            self, start, stop, step=None, *,
+            length: Optional[Integral] = None,
+            include_stop: bool = True
+    ):
+        """Create a new ``UniformGrid``.
+
+        :param start:
+            First point in the grid
+        :param stop:
+            Last point in the grid (some caveats apply, see ``include_stop``)
+        :param step:
+            Desired step size of the grid. Mutually exclusive with ``length``.
+        :param length:
+            Desired length of the grid. Mutually exclusive with ``step``.
+        :param include_stop:
+            Controls the behaviour when ``step`` is not ``None`` and
+            ``stop - start`` is not an integer multiple of ``step``.
+            If ``True`` (default), the actual endpoint of the grid will be
```

Comment:
Changed to
```
        :param always_include_stop:
            Controls the behaviour when ``step`` is not ``None``.
            If ``True`` (default), the endpoint is the smallest
            ``start + k * step``, with ``k`` integer, which is
            greater than or equal to ``stop``.
            Otherwise, the endpoint is the largest
            ``start + k * step``, with ``k`` integer, which is
            smaller than or equal to ``stop``.
```